posix DAEMON=1 build failed on macOS

**Describe the bug**
Build on macOS failed with DAEMON=1.

**To Reproduce**


1. Git commit id: 0cf84da1006d70ac0a3a817406d702a1132eab63
2. IEEE 802.15.4 hardware platform: macOS, posix
3. Build steps
```bash
./bootstrap
make -f src/posix/Makefile-posix DAEMON=1
```

**Expected behavior**
Build successfully.

**Console/log output**
```bash
  CXX      libopenthread_posix_a-udp.o
  CXX      libopenthread_posix_a-virtual_time.o
/Users/simonlin/src/openthread/src/posix/../../src/posix/platform/udp.cpp:86:28: error: variable length arrays are a C99 feature [-Werror,-Wvla-extension]
                    control[OT_APPLE_IGNORE_GNU_FOLDING_CONSTANT(CMSG_SPACE(sizeof(struct in6_pktinfo)) + CMSG_SPACE(sizeof(int)))];
                           ^
/Users/simonlin/src/openthread/src/posix/../../src/posix/platform/udp.cpp:113:26: error: implicit conversion loses integer precision: 'unsigned long' to 'socklen_t' (aka 'unsigned int')
      [-Werror,-Wshorten-64-to-32]
    msg.msg_controllen = sizeof(control);
                       ~ ^~~~~~~~~~~~~~~
2 errors generated.
```

--- a/src/posix/platform/udp.cpp
+++ b/src/posix/platform/udp.cpp
@@ -80,6 +80,7 @@ static bool IsMulticast(const otIp6Address &aAddress)
 static otError transmitPacket(int aFd, uint8_t *aPayload, uint16_t aLength, const otMessageInfo &aMessageInfo)
 {
 #ifdef __APPLE__
+    // use fixed value for CMSG_SPACE is not a constant expression on macOS
     constexpr size_t kBufferSize = 128;
 #else
     constexpr size_t kBufferSize = CMSG_SPACE(sizeof(struct in6_pktinfo)) + CMSG_SPACE(sizeof(int));

--- a/src/posix/platform/udp.cpp
+++ b/src/posix/platform/udp.cpp
@@ -79,15 +79,19 @@ static bool IsMulticast(const otIp6Address &aAddress)
 
 static otError transmitPacket(int aFd, uint8_t *aPayload, uint16_t aLength, const otMessageInfo &aMessageInfo)
 {
+#ifdef __APPLE__
+    constexpr size_t kBufferSize = 128;
+#else
+    constexpr size_t kBufferSize = CMSG_SPACE(sizeof(struct in6_pktinfo)) + CMSG_SPACE(sizeof(int));
+#endif
     struct sockaddr_in6 peerAddr;
-    uint8_t
-                    control[OT_APPLE_IGNORE_GNU_FOLDING_CONSTANT(CMSG_SPACE(sizeof(struct in6_pktinfo)) + CMSG_SPACE(sizeof(int)))];
-    size_t          controlLength = 0;
-    struct iovec    iov;
-    struct msghdr   msg;
-    struct cmsghdr *cmsg;
-    ssize_t         rval;
-    otError         error = OT_ERROR_NONE;
+    uint8_t             control[kBufferSize];
+    size_t              controlLength = 0;
+    struct iovec        iov;
+    struct msghdr       msg;
+    struct cmsghdr *    cmsg;
+    ssize_t             rval;
+    otError             error = OT_ERROR_NONE;
 
     memset(&peerAddr, 0, sizeof(peerAddr));
     peerAddr.sin6_port   = htons(aMessageInfo.mPeerPort);
@@ -108,7 +112,7 @@ static otError transmitPacket(int aFd, uint8_t *aPayload, uint16_t aLength, cons
     msg.msg_name       = &peerAddr;
     msg.msg_namelen    = sizeof(peerAddr);
     msg.msg_control    = control;
-    msg.msg_controllen = sizeof(control);
+    msg.msg_controllen = static_cast<decltype(msg.msg_controllen)>(sizeof(control));
     msg.msg_iov        = &iov;
     msg.msg_iovlen     = 1;
     msg.msg_flags      = 0;
@@ -149,7 +153,7 @@ static otError transmitPacket(int aFd, uint8_t *aPayload, uint16_t aLength, cons
 #ifdef __APPLE__
     msg.msg_controllen = static_cast<socklen_t>(controlLength);
 #else
-    msg.msg_controllen = controlLength;
+    msg.msg_controllen           = controlLength;
 #endif
 
     rval = sendmsg(aFd, &msg, 0);
@@ -299,7 +303,7 @@ otError otPlatUdpBindToNetif(otUdpSocket *aUdpSocket, otNetifIdentifier aNetifId
     {
     case OT_NETIF_UNSPECIFIED:
     {
-#if __linux__
+#ifdef __linux__
         VerifyOrExit(setsockopt(fd, SOL_SOCKET, SO_BINDTODEVICE, nullptr, 0) == 0, error = OT_ERROR_FAILED);
 #else  // __NetBSD__ || __FreeBSD__ || __APPLE__
         unsigned int netifIndex = 0;
@@ -310,7 +314,7 @@ otError otPlatUdpBindToNetif(otUdpSocket *aUdpSocket, otNetifIdentifier aNetifId
     }
     case OT_NETIF_THREAD:
     {
-#if __linux__
+#ifdef __linux__
         VerifyOrExit(setsockopt(fd, SOL_SOCKET, SO_BINDTODEVICE, &gNetifName, strlen(gNetifName)) == 0,
                      error = OT_ERROR_FAILED);
 #else  // __NetBSD__ || __FreeBSD__ || __APPLE__

